File: multeemux.py
#!/usr/bin/env python
import os
import sys
import socket
import random
import argparse
import select
import threading

import signal

signal.signal(signal.SIGINT, lambda x, y: sys.exit(1))

BUF_SIZE = 4096
BIND_HOST = '0.0.0.0'

def usage(p):
    p.print_help()
    sys.exit(1)

def tmux_socket():
    tmux = os.getenv('TMUX')
    if not tmux: raise Exception("Not running tmux")
    return tmux.split(',')[0]

def proxy_data(sockets):
    while True:
        r, w, e = select.select(sockets, [], [], 0)
        for sock in r:
            other = filter(lambda x: x != sock, sockets)[0]
            data = sock.recv(BUF_SIZE)
            if not data:
                sys.exit(0)
            other.sendall(data)

def listen():
    listener = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    tmux = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)

    tmux.connect(tmux_socket())
    for i in range(100): # Plenty of attempts
        port = random.randint(32767, 65535)
        try:
            listener.bind((BIND_HOST, port))
            break
        except socket.error:
            continue
    else:
        raise Exception("Couldn't find a port to bind to ")
    print("Bound to %s:%d" % (BIND_HOST, port))

    listener.listen(1)
    (client, addr) = listener.accept()

    proxy_data((client, tmux))

def connect(hostspec):
    host, port = hostspec.split(":")
    port = int(port)

    remote = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    local = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)

    # Skeeeetchy
    local_socket_name = os.tempnam()
    local.bind(local_socket_name)

    remote.connect((host, port))

    local.listen(1)
    # print
    # Preeetty much the reason we can't have nice things
    threading.Thread(target=lambda: os.system("TMUX=%s tmux new" % (local_socket_name))).run()
    (conn, addr) = local.accept()

    proxy_data((conn, remote))

def _get_parser():
    p = argparse.ArgumentParser()
    p.add_argument("--listen", dest='listen', action="store_true")
    p.add_argument("--connect", dest='connect', action="store")
    return p

if __name__ == '__main__':
    parser = _get_parser()
    args = parser.parse_args()
    if (args.listen and args.connect):
        usage(parser)

    if args.listen:
        listen()
    elif args.connect:
        connect(args.connect)
    else:
        usage()

File: multeemux.rs
use std::os;
use std::io::net::ip::SocketAddr;
use std::io::net::tcp::TcpStream;
use std::io::net::unix::UnixStream;

fn tmux_socket() -> UnixStream {
    let socket_path = os::getenv("TMUX").unwrap();
    match UnixStream::connect(&socket_path.as_slice()) {
        Ok(s) => s,
        Err(e) => fail!("Couldn't connect to tmux: {}", e)
    }
}

fn usage() {
    let args = os::args();
    println!("Usage: {} [--listen] | [-connect HOST:PORT]", args[0]);
    os::set_exit_status(1);
}

fn listen() {
    println!("LISTEN");
}

fn connect(hostspec: &str) {
    let addr: SocketAddr = match from_str(hostspec) {
        Some(h) => h,
        None => fail!("Couldn't parse address: {}", hostspec)
    };
    let mut socket = match TcpStream::connect(addr) {
        Ok(s) => s,
        Err(e) => fail!("Couldn't connect: {}: {}", addr, e)
    };

    let tmux = tmux_socket();

    spawn(proc() {
        println!("Proxy code goes here");

    });

    println!("sleeping forever");
    std::io::timer::sleep(5);
}

fn main() {
    let args = os::args();

    if args.len() == 2 && args[1].eq(&("--listen".to_owned())) {
        return listen();
    }

    if args.len() == 3 && args[1].eq(&("--connect".to_owned())) {
        return connect(args[2]);
    }

    usage();
}
